tests/providers/conftest.py

```diff
@@ -0,0 +1,75 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import pytest
+
+from tests.test_utils import db
+
+# Providers with subpackages
+INNER_PROVIDERS = {
```

Comment:
Should/can we generate this list automatically?
Hard coding has risk that we will forget to update the list when needed

Reply:
We can check for presence of provider.yaml. Or better - rely on `generated/provider_dependencies.json`  instead. the provider_dependencies.json is guaranteed to be updated by pre-commit and it is far easier to use than scanning folders, as well as it's intention is to remain stable when we refactor and move the providers.

Reply:
Another comment here @Taragolis . Maybe we do not need that at all. Don't we want to do it for ALL packages that are "leaves" - not only for providers? I thin that would be a nice compromise between isolation and speed of execution

Reply:
There is also one side-effect that we should take into account here.  This is fine when we run the tests in CI, but the way it is implemented now, it will also run the auto-fixture even if you run a single unit test. This will dramatically slow-down any iterations you do I am afraid (and it does not really matter if we do it only for Providers or for whole airflow)..

Reply:
I think we could just need to scan directory until found one of the folder: `hooks`, `operators`, `sensors`

Reply:
> but the way it is implemented now, it will also run the auto-fixture even if you run a single unit test. This will dramatically slow-down any iterations you do I am afraid (and it does not really matter if we do it only for Providers or for whole airflow).

A lot of tests right now clear some part of DB between the run, so I can't say it dramatically affect of single test.


Single test with fixture (3 runs) on Postgres in breeze
```console

1 passed in 1.95s
1 passed in 1.96s
1 passed in 2.00s
```

Single test without fixture (3 runs) on Postgres in breeze
```console

1 passed in 1.02s
1 passed in 0.98s
1 passed in 0.99s
```

So one second per run it not a big deal and if developer run all tests in single module or just one test class, it would be even unnoticeable.

> Don't we want to do it for ALL packages that are "leaves" - not only for providers (including all airflow packages)?

Initial problem happen because different providers packages could run separately for each other, e.g. SSH and Oracle and if developer do some stuff wrong we could find it only in main.

All others tests run all together, so if one tests from Core affect another then most probably we notice this in Core section

Reply:
> So one second per run it not a big deal and if developer run all tests in single module or just one test class, it would be even unnoticeable.

One second is a lot for tests that do not require the DB at all - especially if you run them from IDE with re-running single tests after the whole suite is run,  The difference is - immediate green/red vs. waiting a second. I often forget to remove `--with-db-init` when I add it and this happens only for one run - the difference is really noticeable.

Compare those two: 

With db reset:

https://user-images.githubusercontent.com/595491/209865511-d2815c9f-0747-4445-adbe-0318705fbc1d.mov

Without:

https://user-images.githubusercontent.com/595491/209865517-0931689a-3491-43ec-824d-8db14100a775.mov

> All others tests run all together, so if one tests from Core affect another then most probably we notice this in Core section

But the same kind of problem happened on multiple occasions for all kinds of tests, not only between providers. The mechanism is the same - data created in one test (in unrelated package) might impact tests from another package. So I think if we attempt to do such isolation, it should be done across the board.

Reply:
1 second difference with this approach only in breeze in MacOS, in native environment differences even less

```console

1 passed in 0.21s
1 passed in 0.23s
1 passed in 0.24s
```

```console

1 passed in 0.03s
1 passed in 0.03s
1 passed in 0.03s
```

Reply:
See the videos above (that's the native env).. It is noticeable, a lot. I hate having to wait even 0.5 to run single test with Ctrl+R. Plus some people have auto-rerrunable tests (there are plugins that run tests automatically after saving). We cannot make that experience so much worse.

Reply:
> But the same kind of problem happened on multiple occasions for all kinds of tests, not only between providers. The mechanism is the same - data created in one test (in unrelated package) might impact tests from another package. So I think if we attempt to do such isolation, it should be done across the board.

Then I suggest to do in a different way. Because such side effect outside of Providers could happened only when two PR merged and create this stuff

1. Migrate all test to pytest (there is less then 100 modules remaining)
2. Create separate pytest plugin with fixtures/context managers
3. Migrate all current tests to use this fixtures, most of them already exists in [conftest.py](https://github.com/apache/airflow/blob/main/tests/conftest.py)
4. Restrict manually create DAGs, TIs and etc and allow only by fixtures/context managers (more tricky part)
5. All core functionality should have setup and teardown which cleanup (most of them already have)

Reply:
> See the videos above (that's the native env).. It is noticeable, a lot.

I think this happen because Init DB required load Flask and their dependencies.
All other tests right not use reset db / init db for cleanup their states. Instead of just run direct queries by this module: https://github.com/apache/airflow/blob/main/tests/test_utils/db.py which run pretty fast

I agree that the current solution a bit dumb but the main target right now reduce chance of side-effects between different providers tests, and the cost about 0.2 sec on single test or +0.2 (plus additional for not expensive logic) on single provider package.

A lot of chance that this solution do not required as soon as we change structure of providers tests.

For all tests we should consider to use different approach

Reply:
> I hate having to wait even 0.5 s to run single test with Ctrl+R. Plus some people have auto-rerrunable tests (there are plugins that run tests automatically after saving). We cannot make that experience so much worse.

This could be nice debate for many reason.

First of all `pytest` itself started for about 1 second with load all plugins plus outputs might take a lot of resources. 

```console
❯ time pytest --help 
...
pytest --help  1.21s user 0.21s system 76% cpu 1.851 total

❯ time pytest --help
...
pytest --help  0.58s user 0.11s system 98% cpu 0.697 total

❯ time pytest --help
...
pytest --help  0.59s user 0.12s system 87% cpu 0.816 total
```

PyCharm also could add latency. Hopefully we not enable any coverage in adopts because PyCharm/IDEA could add additional 5-10 seconds per run in this case.

Personally most of the time I spend in debugger when I try to check is it work as expected or try to figure out why it not working. And usually rerun all the module tests just for sure that last change not break anything new.

As well as if we consider **X** it is a time for create something new then time for cover by tests plus fixing become **X * 5** with most of the time siting with hand on the face and told "Why on Earth you don't work as expected" or "This concept is garbage I need to rewrite my code because tests shows me not everything might work well".

But this is a personal as well as all times only related to my setup, someone who has old laptop could additionally wait more than 0.2 second

Reply:
You convinced me.

I just tested your solution (My videos were with --with-db-init) and yes. I agree it's not noticeable in PyCharm/IntelliJ, so you are likely right that Flask import and others add a lot. And yeah. Having DB reset in this case for every test is quite OK. It is actually likely that those flask imports cause the constraint problems anyway.

Ttry to notice which one is which (it's visible, but barely).

https://user-images.githubusercontent.com/595491/209874550-732ee92c-43dc-4fc1-b3f7-f0b365b0d281.mov

https://user-images.githubusercontent.com/595491/209874540-7a7e5611-d888-45a8-b8fd-02f8df147385.mov

And it has the nice benefit that it will help the users in many situations where they run single tests as well. This overhead is much smallee that I was used to from `--with-db-init` 

However, I think few changes are needed:

1) have a "clear_all()" in the test_utils.db and using it instead of having the list of methods in conftest.py  - this will help with future maintainability

2) What's even more ... I **really** think we should apply it to all tests and completely apply it to the main conftest.py and do it for all modules - I think with such low overhead we can kill few birds with the same stone and the overall overhead will be miinimal. I just did this in main "conftest.py". I don't think we need to detect any providers at all and run it for **everything**

```
@pytest.fixture(scope="module", autouse=True)
def _clear_db_between_modules(request):
    from tests.test_utils import db

    """Clear DB between each separate provider package test runs."""
    print("!!!!!!!!Clearing DB!!!!!!!!")
    db.clear_db_runs()
    db.clear_db_datasets()
    db.clear_db_dags()
    db.clear_db_serialized_dags()
    db.clear_db_sla_miss()
    db.clear_db_pools()
    db.clear_db_connections()
    db.clear_db_variables()
    db.clear_db_dag_code()
    db.clear_db_callbacks()
    db.clear_rendered_ti_fields()
    db.clear_db_import_errors()
    db.clear_db_dag_warnings()
    db.clear_db_xcom()
    db.clear_db_logs()
    db.clear_db_jobs()
    db.clear_db_task_fail()
    db.clear_db_task_reschedule()
    db.clear_dag_specific_permissions()
    db.create_default_connections()
    db.set_default_pool_slots(128)
    yield
```

And run it with -s 

```
pytest -s tests/cli/ tests/core/
```

And it looks like it works as expected:

```
tests/cli/commands/test_info_command.py::TestPiiAnonymizer::test_should_remove_pii_from_path !!!!!!!!Clearing DB!!!!!!!!
PASSED
tests/cli/commands/test_info_command.py::TestPiiAnonymizer::test_should_remove_pii_from_url[postgresql+psycopg2://postgres:airflow@postgres/airflow-postgresql+psycopg2://p...s:PASSWORD@postgres/airflow] PASSED
tests/cli/commands/test_info_command.py::TestPiiAnonymizer::test_should_remove_pii_from_url[postgresql+psycopg2://postgres@postgres/airflow-postgresql+psycopg2://p...s@postgres/airflow] PASSED
tests/cli/commands/test_info_command.py::TestPiiAnonymizer::test_should_remove_pii_from_url[postgresql+psycopg2://:airflow@postgres/airflow-postgresql+psycopg2://:PASSWORD@postgres/airflow] PASSED
tests/cli/commands/test_info_command.py::TestPiiAnonymizer::test_should_remove_pii_from_url[postgresql+psycopg2://postgres/airflow-postgresql+psycopg2://postgres/airflow] PASSED
tests/cli/commands/test_info_command.py::TestAirflowInfo::test_airflow_info PASSED
tests/cli/commands/test_info_command.py::TestAirflowInfo::test_system_info PASSED
tests/cli/commands/test_info_command.py::TestAirflowInfo::test_paths_info PASSED
tests/cli/commands/test_info_command.py::TestAirflowInfo::test_tools_info PASSED
tests/cli/commands/test_info_command.py::TestAirflowInfo::test_show_info PASSED
tests/cli/commands/test_info_command.py::TestAirflowInfo::test_show_info_anonymize PASSED
tests/cli/commands/test_info_command.py::TestInfoCommandMockHttpx::test_show_info_anonymize_fileio PASSED
tests/cli/commands/test_jobs_command.py::TestCliConfigList::test_should_report_success_for_one_working_scheduler !!!!!!!!Clearing DB!!!!!!!!
PASSED
```

Reply:
I think it costs us almost nothing and is a **long term** solution if we do it this way.

Reply:
I could check also `request` fixture tomorrow, I thought it should contain session information, such as how many tests collected and if it equal (or less, I don't know how it possible) `1` then do not cleanup anything.

Reply:
And also remove `yield` in this fixture we do not need any teardown, at least yet

Reply:
I could check also request fixture tomorrow, I thought it should contain session information, such as how many tests collected and if it equal (or less, I don't know how it possible) 1 then do not cleanup anything.

> I think this is not really needed now - the overhead with using the test_utils classes is really low indeed. I believe we should enable it for all tests and compare execution time on CI and if it is < 5% overhead in total or so, we can leave it on for all modules.

Reply:
For long term we also need to consider get rid of most of the stuff in `test.test_utils` and replace them by equivalent as fixture/context manager/markers but this better complete after migrate remaining tests to `pytest` and document them.

Because most of the new contributors just copy-paste some stuff from one test to another without any idea how it actually work.

Reply:
> For long term we also need to consider get rid of most of the stuff in `test.test_utils` and replace them by equivalent as fixture/context manager/markers but this better complete after migrate remaining tests to `pytest` and document them.
> 
> Because most of the new contributors just copy-paste some stuff from one test to another without any idea how it actually work.

Yep. Full agreement here.